File: commons-parsers/src/main/java/xyz/wagyourtail/commons/parsers/impl/json5/Json5StringConstant.java
package xyz.wagyourtail.commons.parsers.impl.json5;

import lombok.val;
import xyz.wagyourtail.commons.core.StringUtils;
import xyz.wagyourtail.commons.core.reader.CharReader;
import xyz.wagyourtail.commons.core.reader.StringCharReader;
import xyz.wagyourtail.commons.parsers.Data;
import xyz.wagyourtail.commons.parsers.StringData;

public class Json5StringConstant extends StringData.OnlyRaw<Data.SingleContent<String>> {


    private Json5StringConstant(String content) {
        super(content, Json5StringConstant::getContentChecked);
    }

    public Json5StringConstant(CharReader<?> content) {
        super(content, Json5StringConstant::getContentChecked);
    }

    public static Json5StringConstant parse(String rawContent) {
        StringCharReader reader = new StringCharReader(rawContent);
        val number = new Json5StringConstant(reader);
        reader.expectEOS();
        return number;
    }

    public static Json5StringConstant unchecked(String rawContent) {
        return new Json5StringConstant(rawContent);
    }

    private static SingleContent<String> getContentChecked(CharReader<?> reader) {
        int quote = reader.peek();
        if (quote == '"' || quote == '\'') {
            return new SingleContent<>(
                    reader.takeString(CharReader.TAKE_STRING_NO_TRANSLATE_ESCAPES | CharReader.TAKE_STRING_LEINIENT | CharReader.TAKE_STRING_ESCAPE_NEWLINE, String.valueOf((char) quote))
            );
        }
        throw reader.createException("Expected string start char but got: " + (char) quote);
    }

    public String getValue() {
        val content = getRawContent();
        return StringUtils.translateEscapes(content.substring(1, content.length() - 1).replace("\\\n", ""));
    }

}

File: commons-parsers-kt-kt/src/commonMain/kotlin/xyz/wagyourtail/commonskt/parsers/impl/json5/Json5StringConstant.kt
package xyz.wagyourtail.commonskt.parsers.impl.json5

import xyz.wagyourtail.commonskt.parsers.Data
import xyz.wagyourtail.commonskt.parsers.StringData
import xyz.wagyourtail.commonskt.reader.CharReader
import xyz.wagyourtail.commonskt.utils.translateEscapes

class Json5StringConstant(rawContent: String) : StringData.OnlyRaw<Data.SingleContent<String>>(rawContent) {

    constructor(content: SingleContent<String>) : this(content.toString())

    companion object : StringDataBuilder<Json5StringConstant, Data.SingleContent<String>> {

        override fun invoke(rawContent: CharReader<*>): Json5StringConstant {
            return Json5StringConstant(checkedBuildContent(rawContent))
        }

        override fun checkedBuildContent(reader: CharReader<*>): SingleContent<String> {
            val quote = reader.peek()
            if (quote == '\"' || quote == '\'') {
                return SingleContent(
                    reader.takeString(
                        leinient = true,
                        escapeNewline = true,
                        noTranslateEscapes = true,
                        quote = quote.toString()
                    )
                )
            }
            throw reader.createException("Expected string start char but got: $quote")
        }

    }

    val value: String
        get() {
            val content = rawContent
            return content.substring(1, content.length - 1).replace("\\\n", "").translateEscapes()
        }

    override fun checkedBuildContent(reader: CharReader<*>): SingleContent<String> {
        return Companion.checkedBuildContent(reader)
    }

}